README.md
Digital Marketplace API client
=========================

![Python 3.9](https://img.shields.io/badge/python-3.9-blue.svg)

## What's in here?

API clients for Digital Marketplace [Data API](https://github.com/Crown-Commercial-Service/digitalmarketplace-api) and
[Search API](https://github.com/Crown-Commercial-Service/digitalmarketplace-search-api).

Originally was part of [Digital Marketplace Utils](https://github.com/Crown-Commercial-Service/digitalmarketplace-utils).


## Running the tests

Install Python dependencies:

```
make bootstrap
invoke requirements-dev
```

Run the tests:

```
invoke test
```

## Usage examples

```python

data_client = apiclient.DataAPIClient(api_url, api_access_token)
services = data_client.find_services_iter(framework=frameworks)

```

## Releasing a new version

To update the package version, edit the `__version__ = ...` string in `dmapiclient/__init__.py`,
commit and push the change and wait for CI to create a new version tag.

Once the tag is available on GitHub, the new version can be used by the apps by adding the following
line to the app `requirements.txt` (replacing `X.Y.Z` with the current version number):

```
git+https://github.com/Crown-Commercial-Service/digitalmarketplace-apiclient.git@X.Y.Z#egg=digitalmarketplace-apiclient==X.Y.Z
```

When changing a major version number consider adding a record to the `CHANGELOG.md` with a
description of the change and an example of the upgrade process for the client apps.

## Licence

Unless stated otherwise, the codebase is released under [the MIT License][mit].
This covers both the codebase and any sample code in the documentation.

The documentation is [&copy; Crown copyright][copyright] and available under the terms
of the [Open Government 3.0][ogl] licence.

[mit]: LICENCE
[copyright]: http://www.nationalarchives.gov.uk/information-management/re-using-public-sector-information/uk-government-licensing-framework/crown-copyright/
[ogl]: http://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/


CHANGELOG.md
# Digital Marketplace API clients changelog

Records breaking changes from major version bumps

## 24.0.0
Update to use python 3.9 and drop support for python 3.8.

Updated other packages so minimum Flask requirement is now v2

## 23.0.0
Update to use python 3.8 as python 3.6 is reaching end of life.

Dropping support for:
- python 3.6
- python 3.7

## 22.0.0

Mark this package as PEP 561 compatible. If you have dmapiclient as a dependency and run type checking, the type checker
will now check that you're using dmapiclient correctly. This might break your application's type checking if you're 
using dmapiclient incorrectly.

## 21.0.0

All "attributes" of apiclient instances have been made private and replaced with read-only properties of the same name.
This is to encourage use of apiclients as objects that are unmutated after construction, making their use (hopefully)
threadsafe.

The attributes in question comprise `RETRIES`, `RETRIES_BACKOFF_FACTOR`, `RETRIES_FORCE_STATUS_CODES`, `base_url`,
`auth_token`, `enabled`, `timeout`. It seems unlikely that any code was mutating these values
post-construction/`init_app`, but it's worth doing a check before upgrading.

## 20.0.0

PR: [#205](https://github.com/alphagov/digitalmarketplace-apiclient/pull/205)

The with_declarations parameter of find_framework_suppliers method of DataAPIClient now defaults to True instead of None. This will increase significantly the size of all replies using it. 

## 19.0.0

PR: [#152](https://github.com/alphagov/digitalmarketplace-apiclient/pull/152)

We're now using urllib3's Retry util to retry failed requests made by the base api client. This has allowed us to
remove backoff from `make_iter_method` and so also remove the package from the api client's dependencies. It also means
that we no longer need `HTTPTemporaryError` so it's been removed.

Anything that uses the api client and has it's own backoff setup (such as some of our scripts)
will need to adjust (or remove) their config to prevent potentially making far too many retry requests.

Anything that uses `HTTPTemporaryError` will need to remove it. The only place I've found it is in a few one off scripts
in the scripts repo for backoff setup.

## 18.0.0

PR: [#151](https://github.com/alphagov/digitalmarketplace-apiclient/pull/151)

The default request error message for unrecognised/unhandled errors has changed from 'Request failed' to 'Unknown request failure in dmapiclient'. 'Request failed' should be seen only in exceptional circumstances - most exceptions raised from failed requests should now hold the str() and repr() calls of the request error as their message.

Any code that relies on the specific format of the request errors raised by the apiclient will need to be updated.

## 17.0.0

PR: [#143](https://github.com/alphagov/digitalmarketplace-apiclient/pull/143)

No longer python2 compatible.

## 16.0.0

PR: [#137](https://github.com/alphagov/digitalmarketplace-apiclient/pull/137)

Moves `api_stubs.py` from this repo (digitalmarketplace-apiclient) to utils (digitalmarketplace-utils). Any repositories using
the stubs need to now use stubs from dm-utils.

Ensure that you update frontend app requirements files to use dmutils>=36.6.0 and that you update imports. The interface for
some of the stubs has changed slightly, and more detailed responses are returned for the frameworks stub, so be aware
of this as you migrate.

Old:
```python
import dmapiclient.api_stubs
```

New:
```python
import dmutils.api_stubs
```


## (14.5.0 ->) 15.0.0

:rotating_light: **Note:** there was mistaken versioning between `14.5.0` and `15.0.0`, meaning that the changes
described below for `15.0.0` actually applied from `14.5.0` onwards.

PR: [#123](https://github.com/alphagov/digitalmarketplace-apiclient/pull/123)

Removes `get_direct_award_project_services_iter` in favour of `find_direct_award_project_services_iter`, which is the
more consistent naming scheme.

Old:
```python
data_api_client.get_direct_award_project_services_iter(...)
```

New:
```python
data_api_client.find_direct_award_project_services_iter(...)
```

## 14.0.0

PR: [#116](https://github.com/alphagov/digitalmarketplace-apiclient/pull/116)

`search_services` method on the search-api-client has been generalized to `search` and now takes a `doc_type` argument.
This allows more than just services to be searched.

`aggregate` method on the search-api-client has been generalized to `aggregate` to allow for aggregating
briefs too. It takes an extra argument, `doc_type`.

Old:
```
search_api_client.aggregate_services(index=index, q=q, aggregations=aggregations)
search_api_client.search_services(index, q=q, page=page, id_only=id_only, **filters)
```

New:
```
search_api_client.aggregate(index=index, doc_type='services', q=q, aggregations=aggregations)
search_api_client.search(index, doc_type, q=q, page=page, id_only=id_only, **filters)
```

## 13.0.0

PR: [#115](https://github.com/alphagov/digitalmarketplace-apiclient/pull/115)

`name` parameter has been added to `update_user` method,  so any code calling
this method with positional parameters needs to be updated to use this version.

## 12.0.0

PR: [#111](https://github.com/alphagov/digitalmarketplace-apiclient/pull/111)

`mapping` parameter is now required for Search API index creation, as we must be
able to distinguish between the 'services' mapping and the forthcoming potential
`briefs` mapping. We might also in future have multiple mappings to cover a new
non-compatible G-Cloud index, for advance preparation of the new index in advance
of the framework going live.

## 11.0.0

PR: [#105](https://github.com/alphagov/digitalmarketplace-apiclient/pull/105)

Gives `user_id` parameter a default for `find_direct_award_project_searches` and `get_direct_award_project_services`,
which changes the order of their positional parameters.

The old/new examples below are calling the methods with no `user_id` and a `project_id` of 123.

Old
```

dmapiclient/api_stubs.py

def framework(status="open", slug="g-cloud-7", name=None, clarification_questions_open=True, lots=None):
    if slug == "g-cloud-7":
        name = "G-Cloud 7"
    elif slug == "digital-outcomes-and-specialists":
        name = "Digital Outcomes and Specialists"
    elif slug == "digital-outcomes-and-specialists-2":
        name = "Digital Outcomes and Specialists 2"
    else:
        name = slug.replace("-", " ").title()

    lots = lots or []

    return {
        "frameworks": {
            "status": status,
            "clarificationQuestionsOpen": clarification_questions_open,
            "name": name,
            "slug": slug,
            "lots": lots,
        }
    }


def lot(slug="some-lot", allows_brief=False, one_service_limit=False):
    return {
        "id": 1,
        "slug": slug,
        "name": slug.replace("-", " ").title(),
        "allowsBrief": allows_brief,
        "oneServiceLimit": one_service_limit,
    }


def brief(status="draft",
          framework_slug="digital-outcomes-and-specialists",
          lot_slug="digital-specialists",
          user_id=123,
          framework_name="Digital Outcomes and Specialists",
          framework_framework="digital-outcomes-and-specialists",
          clarification_questions=None,
          clarification_questions_closed=False):
    brief = {
        "briefs": {
            "id": 1234,
            "title": "I need a thing to do a thing",
            "frameworkSlug": framework_slug,
            "frameworkName": framework_name,
            "frameworkFramework": framework_framework,
            "lotSlug": lot_slug,
            "status": status,
            "users": [{"active": True,
                       "role": "buyer",
                       "emailAddress": "buyer@email.com",
                       "id": user_id,
                       "name": "Buyer User"}],
            "createdAt": "2016-03-29T10:11:12.000000Z",
            "updatedAt": "2016-03-29T10:11:13.000000Z",
            "clarificationQuestions": clarification_questions or [],
        }
    }
    if status in ("live", "closed"):
        brief['briefs']['publishedAt'] = "2016-03-29T10:11:14.000000Z"
        brief['briefs']['applicationsClosedAt'] = "2016-04-07T00:00:00.000000Z"
        brief['briefs']['clarificationQuestionsClosedAt'] = "2016-04-02T00:00:00.000000Z"
        brief['briefs']['clarificationQuestionsAreClosed'] = clarification_questions_closed

    return brief


def supplier(id=1234, contact_id=4321, other_company_registration_number=0, company_details_confirmed=True):
    data = {
        "suppliers": {
            "companiesHouseNumber": "12345678",
            "companyDetailsConfirmed": company_details_confirmed,
            "contactInformation": [
                {
                    "address1": "123 Fake Road",
                    "city": "Madeupolis",
                    "contactName": "Mr E Man",
                    "email": "mre@company.com",
                    "id": contact_id,
                    "links": {
                        "self": "http://localhost:5000/suppliers/{id}/contact-information/{contact_id}".format(
                            id=id, contact_id=contact_id
                        )
                    },
                    "phoneNumber": "01234123123",
                    "postcode": "A11 1AA",
                    "website": "https://www.mre.company"
                }
            ],
            "description": "I'm a supplier.",
            "dunsNumber": "123456789",
            "id": id,
            "links": {
                "self": "http://localhost:5000/suppliers/{id}".format(id=id)
            },
            "name": "My Little Company",
            "organisationSize": "micro",
            "registeredName": "My Little Registered Company",
            "registrationCountry": "country:GB",
            "service_counts": {
                "G-Cloud 9": 109,
                "G-Cloud 8": 108,
                "G-Cloud 7": 107,
                "G-Cloud 6": 106,
                "G-Cloud 5": 105,
            },
            "tradingStatus": "limited company",
            "vatNumber": "111222333"
        }
    }

    if other_company_registration_number:
        data['suppliers']['otherCompanyRegistrationNumber'] = other_company_registration_number
        # We allow one or other of these registration numbers, but not both
        del data['suppliers']['companiesHouseNumber']
        # Companies without a Companies House number aren't necessarily overseas, but they might well be
        data['suppliers']['registrationCountry'] = 'country:NZ'

    return data
